### PR TITLE
l4proxy: Read from all upstream connections before fully closing them

### LINKED ISSUE #40: l4proxy: Upstream connections may be closed before reading from them
## Problem

The close of upstream connections:

https://github.com/mholt/caddy-l4/blob/31d74d38aa3a41c1563f25dfbee24ced3d4afae2/modules/l4proxy/proxy.go#L173-L178

may happen before the reading of replies from the upstream connections:

https://github.com/mholt/caddy-l4/blob/31d74d38aa3a41c1563f25dfbee24ced3d4afae2/modules/l4proxy/proxy.go#L256-L263

As a result, **the replies from upstream servers will never be forwarded to the downstream clients**. 

## Reproducer

Given a slow Echo server:

```go
package main

import (
	"io"
	"log"
	"net"
	"time"
)

type SlowConn struct {
	net.Conn
}

func (c SlowConn) Write(p []byte) (n int, err error) {
	time.Sleep(5 * time.Millisecond) // Write slowly
	n, err = c.Conn.Write(p)
	return
}

func main() {
	ln, err := net.Listen("tcp", ":7070")
	if err != nil {
		log.Panicln(err)
	}
	defer ln.Close()

	for {
		conn, err := ln.Accept()
		if err != nil {
			log.Panicln(err)
		}

		go handle(conn)
	}
}

func handle(conn net.Conn) {
	c := SlowConn{Conn: conn}
	if _, err := io.Copy(c, c); err != nil {
		log.Printf("err: %v\n", err)
	}
	conn.Close()
}
```

Everything is OK:

```bash
$ echo "hi" | nc localhost 7070
hi
```

Run Caddy with the following JSON config:

```json
{                                       
        "apps": {                       
                "layer4": {                     
                        "servers": {                    
                                "srv0": {
                                        "listen": [":8080"],
                                        "routes": [                             
                                                {                       
                                                        "handle": [     
                                                                {
                                                                        "handler": "proxy",
                                                                        "upstreams": [
                                                                                {"dial": ["localhost:7070"]}
                                                                        ]
                                                                }
                                                        ]
                                                }
                                        ]
                                }
                        }
                }
        }
}
```

Nothing will be displayed:

```bash
$ echo "hi" | nc localhost 8080
```

### PR DESCRIPTION
**Pull request by RussellLuo**
This closes #40.

### CODE REVIEW

**Review by mholt — APPROVED** [review]
Nice, this is much more elegant. And thank you for diving in and figuring out a way to unblock the write-to side! Thanks for the contribution.

(Would you like push access to the repo? Then you can contribute in a branch instead of a fork if you want to.)

**Comment by RussellLuo** [comment]
> (Would you like push access to the repo? Then you can contribute in a branch instead of a fork if you want to.)

Sure, it's my pleasure! Thanks for your trust.

**Pull request merged**
